File: src/backyard_flyer/planner.py
# Waypoint Planner (the 10m box)

# Purpose: generate waypoints relative to the current position

# from __future__ import annotations

from dataclasses import dataclass
from typing import List

from .config import MissionConfig
from .types import Waypoint

# Why class instead of a function?
    # planner can later support multiple patterns (triangle, lawnmower, etc.)
    # without changing the rest of the code.

@dataclass
class BoxPlanner:
    cfg: MissionConfig

    def build_box(self, origin_north: float, origin_east: float) -> List[Waypoint]:
        s = self.cfg.box_size_m

        return [
            Waypoint(origin_north + s, origin_east + 0.0, self.cfg.target_altitude_m),
            Waypoint(origin_north + s, origin_east + s, self.cfg.target_altitude_m),
            Waypoint(origin_north + 0.0, origin_east + s, self.cfg.target_altitude_m),
            Waypoint(origin_north + 0.0, origin_east + 0.0, self.cfg.target_altitude_m),
        ]

File: cpp/backyard_flyer_core/src/planner.cpp
#include "backyard/planner.hpp"

namespace backyard {

std::array<NED, 4> BoxPlanner::BuildBox(double n0, double e0) const {
    const double d = -cfg_.target_altitude_m;
    const double s = cfg_.box_size_m;

    return {
        NED{n0 + s, e0 + 0.0, d},
        NED{n0 + s, e0 + s, d},
        NED{n0 + 0.0, e0 + s, d},
        NED{n0 + 0.0, e0 + 0.0, d},
    };
}

} // namespace backyard